### PR TITLE
onboard create env command

### PR DESCRIPTION
**Pull request by boppanasusanth**
# Odin version(x.y.z)

### CODE REVIEW

File: cmd/create/environment.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	},
+}
+
+func init() {
+	environmentCmd.Flags().StringVar(&envName, "name", "", "name of the environment to be created")
+	environmentCmd.Flags().StringVar(&accounts, "account", "", "list of comma separated cloud provider accounts")
+	environmentCmd.Flags().StringVar(&provisioningType, "provisioningType", "", "provisioning type of the environment")
+	err := environmentCmd.MarkFlagRequired("name")
+	if err != nil {
+		log.Println("Error marking 'name' flag as required:", err)
```
Comment (akshaypatidar1999) [inline cmd/create/environment.go:37]: Use `log.Fatal`. This will log as well exit with non zero status code

File: cmd/create/environment.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		execute(cmd)
+	},
+}
+
+func init() {
+	environmentCmd.Flags().StringVar(&envName, "name", "", "name of the environment to be created")
+	environmentCmd.Flags().StringVar(&accounts, "account", "", "list of comma separated cloud provider accounts")
+	environmentCmd.Flags().StringVar(&provisioningType, "provisioningType", "", "provisioning type of the environment")
```
Comment (akshaypatidar1999) [inline cmd/create/environment.go:34]: Change this to `provisioning-type` to be consistent with `list env` command
https://github.com/dream11/odin/blob/2.x.x/cmd/list/environment.go#L36

File: cmd/configure/configure.go [hunk 3]
```diff
@@ -68,7 +68,7 @@ func execute(cmd *cobra.Command) {
 }
 
 func createConfigFileIfNotExist() {
-	dirPath := path.Join(os.Getenv("HOME"), "." + app.App.Name)
+	dirPath := path.Join(os.Getenv("HOME"), "."+app.App.Name)
```
Comment (yash-saxena-d11) [inline cmd/configure/configure.go:71]: why this change? the previous formatting looks better

File: cmd/create/create.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+import (
+	"github.com/dream11/odin/cmd"
+	"github.com/spf13/cobra"
+)
+
+// createCmd represents the create command
+var createCmd = &cobra.Command{
+	Use:   "create",
+	Short: "Create resources",
```
Comment (yash-saxena-d11) [inline cmd/create/create.go:11]: what other resources can be created through this apart from environments?

File: cmd/create/environment.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		return cobra.NoArgs(cmd, args)
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		execute(cmd)
+	},
+}
+
+func init() {
+	environmentCmd.Flags().StringVar(&envName, "name", "", "name of the environment to be created")
+	environmentCmd.Flags().StringVar(&accounts, "account", "", "list of comma separated cloud provider accounts")
```
Comment (akshaypatidar1999) [inline cmd/create/environment.go:33]: Lets rename this to `accounts` since there can be multiple

File: internal/service/environment.go [hunk 6]
```diff
@@ -23,3 +25,33 @@ func (e *Environment) ListEnvironments(ctx *context.Context, request *environmen
...
+		if err != nil {
+			// Check for the end of the stream
+			if err == context.Canceled {
+				break
+			}
+
+			log.Fatalf("Error receiving stream: %v", err)
+		}
+
+		fmt.Println(response)
```
Comment (akshaypatidar1999) [inline internal/service/environment.go:53]: We need a spinner instead of printing response everytime in new line

File: cmd/create/environment.go (hunk 1 above)
Comment (yash-saxena-d11) [inline cmd/create/environment.go:37]: can't log.fatal suffice?

File: internal/service/environment.go [hunk 7]
```diff
@@ -23,3 +25,33 @@ func (e *Environment) ListEnvironments(ctx *context.Context, request *environmen
...
+	log.Info("Creating environment...")
+	for {
+		response, err := stream.Recv()
+		if err != nil {
+			// Check for the end of the stream
+			if err == context.Canceled {
+				break
+			}
+
+			log.Fatalf("Error receiving stream: %v", err)
```
Comment (yash-saxena-d11) [inline internal/service/environment.go:50]: should this go to the cli?

File: internal/service/environment.go (hunk 6 above)
Comment (yash-saxena-d11) [inline internal/service/environment.go:53]: 1. why not use log?
2. should the response go directly to cli?

**Review by yash-saxena-d11 — COMMENTED** [review]
minor feedback and questions

File: cmd/configure/configure.go (hunk 3 above)
Comment (boppanasusanth) [reply cmd/configure/configure.go:71]: yes. the previous one is better. but running `golangci-lint run -E gofmt -E gci --fix` did this.

File: cmd/create/create.go (hunk 4 above)
Comment (boppanasusanth) [reply cmd/create/create.go:11]: label, service-set

File: cmd/create/environment.go (hunk 1 above)
Comment (boppanasusanth) [reply cmd/create/environment.go]: done

File: cmd/create/environment.go (hunk 2 above)
Comment (boppanasusanth) [reply cmd/create/environment.go]: done

File: cmd/create/environment.go (hunk 5 above)
Comment (boppanasusanth) [reply cmd/create/environment.go]: done

File: cmd/create/environment.go (hunk 1 above)
Comment (boppanasusanth) [reply cmd/create/environment.go]: done.

File: internal/service/environment.go (hunk 7 above)
Comment (boppanasusanth) [reply internal/service/environment.go]: changed. returned the err instead of printing in cli

File: internal/service/environment.go (hunk 6 above)
Comment (surajgour-d11) [reply internal/service/environment.go]: done

Comment (surajgour-d11) [reply internal/service/environment.go]: spinner added

File: cmd/create/environment.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		Accounts:         splitProviderAccount(accounts),
+		ProvisioningType: provisioningType,
+	})
+
+	if err != nil {
+		log.Fatal("Failed to create environment ", err)
+	}
+}
+
+func splitProviderAccount(providerAccounts string) []string {
```
Comment (akshaypatidar1999) [inline cmd/create/environment.go:55]: This can be a utility method that takes string and delimiter as input and returns a string array. Move this to pkg

File: internal/service/environment.go [hunk 9]
```diff
@@ -2,13 +2,30 @@ package service
...
+	"github.com/briandowns/spinner"
 	environment "github.com/dream11/odin/proto/gen/go/dream11/od/environment/v1"
+	log "github.com/sirupsen/logrus"
 )
 
 // Environment performs operation on environment like create, list, describe, delete
 type Environment struct{}
 
+// SpinnerColor Defines color of spinner
+var SpinnerColor = "fgHiBlue"
```
Comment (akshaypatidar1999) [inline internal/service/environment.go:18]: Define these variables in constants. They will be reused across all commands

File: proto/gen/go/dream11/od/dto/v1/service_task.pb.go [hunk 10]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.31.0
-// 	protoc        v4.24.4
+// 	protoc-gen-go v1.33.0
+// 	protoc        v3.21.12
```
Comment (akshaypatidar1999) [inline proto/gen/go/dream11/od/dto/v1/service_task.pb.go:4]: Why downgrading protoc version?

File: internal/service/environment.go [hunk 11]
```diff
@@ -23,3 +40,36 @@ func (e *Environment) ListEnvironments(ctx *context.Context, request *environmen
...
+		if err != nil {
+			if err == context.Canceled || err == io.EOF {
+				break
+			}
+			return err
+		}
+		spinner.Prefix = fmt.Sprintf(" %s  ", response.Message)
+		spinner.Start()
+	}
+	log.Info("Environment created successfully")
```
Comment (akshaypatidar1999) [inline internal/service/environment.go:72]: The stream will always exit successfully so this log will be always printed. We need to check the status to know whether the environment was created or not?

File: internal/service/environment.go [hunk 12]
```diff
@@ -23,3 +40,36 @@ func (e *Environment) ListEnvironments(ctx *context.Context, request *environmen
...
+	stream, err := client.CreateEnvironment(*requestCtx, request)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Creating environment...")
+	spinner := spinner.New(spinner.CharSets[SpinnerType], SpinnerDelay)
+	spinner.Color(SpinnerColor, SpinnerStyle)
+
+	for {
```
Comment (akshaypatidar1999) [inline internal/service/environment.go:60]: If multiple accounts are passed then we have a different status for each of the account. So ideally there should be a multi spinner, one for each account

File: internal/service/environment.go [hunk 13]
```diff
@@ -23,3 +40,36 @@ func (e *Environment) ListEnvironments(ctx *context.Context, request *environmen
...
+		response, err := stream.Recv()
```
Comment (akshaypatidar1999) [inline internal/service/environment.go:61]: What happens if `stream.Recv()` does not return anything?

File: internal/service/environment.go [hunk 14]
```diff
@@ -23,3 +40,36 @@ func (e *Environment) ListEnvironments(ctx *context.Context, request *environmen
...
+		spinner.Stop()
+		if err != nil {
```
Comment (akshaypatidar1999) [inline internal/service/environment.go:63]: We should catch the grpc connection closed errors and retry automatically instead of asking user to retry

File: cmd/create/environment.go (hunk 8 above)
Comment (surajgour-d11) [reply cmd/create/environment.go:55]: The method signature you've mentioned is basically `Split`. This method is created to abstract the provider account splitting logic.
Moving this to pkg

File: internal/service/environment.go (hunk 13 above)
Comment (surajgour-d11) [reply internal/service/environment.go:61]: According to doc
```
RecvMsg blocks until it receives a message into m or the stream is
done. It returns io.EOF when the stream completes successfully. On
any other error, the stream is aborted and the error contains the RPC
status.
```
Stream will either return a response, error or EOF

File: internal/service/environment.go (hunk 11 above)
Comment (surajgour-d11) [reply internal/service/environment.go]: Fixed
PS: Stream fails with error whenever deploy throws creation failed error.

File: internal/service/environment.go (hunk 12 above)
Comment (surajgour-d11) [reply internal/service/environment.go:50]: For multiple accounts, env creation is considered successful only when it is created in all the mentioned account. So no need for account specific status

**Review by akshaypatidar1999 — APPROVED** [review]

File: internal/service/environment.go (hunk 14 above)
Comment (surajgour-d11) [reply internal/service/environment.go:53]: Need to identify error types which should be retried automatically.
Will address in separate PR

**Review by yash-saxena-d11 — APPROVED** [review]
